wandb/sdk/wandb_artifacts.py

```diff
@@ -529,6 +529,25 @@ def do_write(f: IO) -> None:
 
         return entry
 
+    def items(self) -> Iterable[Tuple[str, ArtifactEntry]]:
+        if self._logged_artifact:
+            return self._logged_artifact.items()
+
+        def iter():
+            for k, v in self.manifest.entries.items():
```

Comment:
How does iteration over a manifest work? A flat directory of files/folders like walking an os file-tree?

FYI, the Pythonic way to provide iteration would be to use `__iter__` and `__next__`, so that you can write

```python
for file in artifact:
   print(file)
```

There's subtleties with generators, maybe, to keep watch on.

But note: if we ever use `[]` syntax for a `wandb.Dataset` subclass of `Artifact`, we would need to make sure that the iterator and indexing behaviors match. To do otherwise is un-Pythonic and surprising.

Reply:
I was copying the `dict` semantics here, because we want to present these as `(k,v)` tuples where `k` is the path in the artifact and `v` is the entry metadata.

I could definitely implement `__iter__` and `__next__` but I was worried this would mislead users into using `__getitem__` and `__setitem__`, which while implemented have different semantics. They operate on `WBValue`, NOT `ArtifactEntry` types.

I'd love your thoughts on what you think the most natural APIs here are.

If I could start over, I'd make `artifact['path/to/file']` return the `ArtifactEntry` associated with that path. With your operator overloading trick, we could do: `artifact / 'path' / 'to' / 'file'` to achieve this instead.